src/appleseed.python/bindblenderprogressivetilecallback.cpp

```diff
@@ -172,20 +174,112 @@ namespace
+                        "void main()\n"
+                        "{\n"
+                        "   gl_Position = vec4(pos, 1.0);\n"
+                        "   texture_coords = vec3(tex_coords);\n
+                        "}\0";
+          
+                    const char* fragment_shader_source =
+                        "#version 330 core\n"
+                        "out vec4 tex;\n"
+                        "in vec3 texture_coords;\n"
+                        "unform sampler2D texture;\n"
+                        "void main()\n"
+                        "{\n"
+                        "   tex = vec4(texture, texture_coords);\n"
+                        "}\n\0";
+                    
+                    GLuint vertex_shader_id = glCreateShader(GL_VERTEX_SHADER);
+                    glShaderSource(vertex_shader_id, 1, &vertex_shader_source, nullptr);
+                    glCompileShader(vertex_shader_id);
+                    
+                    // Check for vertex_shader compile errors.
+                    int success;
+                    const int length = 512;
+                    char infoLog[length];
+                    glGetShaderiv(vertex_shader_id, GL_COMPILE_STATUS, &success);
+                    if (!success)
+                    {
+                        glGetShaderInfoLog(vertex_shader_id, length, nullptr, infoLog);
+                    }
+                
+                    GLuint fragment_shader_id = glCreateShader(GL_FRAGMENT_SHADER);
+                    glShaderSource(fragment_shader_id, 1, &fragment_shader_source, nullptr);
+                    glCompileShader(fragment_shader_id);
+                    
+                    // Check for fragment_shader compile errors.
+                    glGetShaderiv(fragment_shader_id, GL_COMPILE_STATUS, &success);
+                    if (!success)
+                    {
+                        glGetShaderInfoLog(fragment_shader_id, length, nullptr, infoLog);
+                    }
+                
+                    m_shader_program_id = glCreateProgram();
+                    glAttachShader(m_shader_program_id, vertex_shader_id);
+                    glAttachShader(m_shader_program_id, fragment_shader_id);
+                    glLinkProgram(m_shader_program_id);
+                    
+                    // Check for linking errors.
+                    glGetProgramiv(m_shader_program_id, GL_LINK_STATUS, &success);
+                    if (!success) 
+                    {
+                        glGetProgramInfoLog(m_shader_program_id, length, nullptr, infoLog);
+                    }
+                    
+                    glDeleteShader(vertex_shader_id);
+                    glDeleteShader(fragment_shader_id);
+               
+                    //Vertex and texture coordinates.
+                    GLfloat vertices[] = 
+                    {
+                        x    , y    , 0.0f, 0.0f, 1.0f
+                      , x + w, y    , 0.0f, 1.0f, 1.0f
```

Comment:
Let's say I call `draw_pixels(A, B, C, D)` and `m_updated_data_buffer` is `false`. This mean the GLSL program is compiled and vertices are setted. `m_updated_data_buffer` is now `true`.

Then I call `draw_pixels(E, F, G, H)`. If I follow the code behavior correctly, buffer datas will not be update.

I suggest to separate the compilation GLSL code that can be done only once. And the  buffer update that is required at each `draw_pixels()` call.

Reply:
Thanks, I was not sure earlier if the vertex data will be updated on each call of the draw_pixels() function or not. Thanks for suggestion 👍

Reply:
Well this code should work without any problem